include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -360,11 +364,153 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
       // actually create triangulation
       new_tria->create_triangulation(construction_data);
 
+      // save mesh
+      coarse_grid_triangulations[min_level - 1] = new_tria;
+      temp_tria.coarsen_global();
+    }
+
+    // find out about the number of MPI processes in shared memory regions,
+    // which will later be used to construct a coarser triangulation only on a
+    // single process per MPI rank
+    unsigned int const group_size = [&]() {
+      int rank;
+      MPI_Comm_rank(mpi_comm, &rank);
+
+      MPI_Comm comm_shared;
+      MPI_Comm_split_type(mpi_comm, MPI_COMM_TYPE_SHARED, rank, MPI_INFO_NULL, &comm_shared);
+
+      int size_shared;
+      MPI_Comm_size(comm_shared, &size_shared);
+      MPI_Comm_free(&comm_shared);
+
+      // determine maximum, since some shared memory communicators
+      // might not be filed completely
+      return Utilities::MPI::max(size_shared, mpi_comm);
+    }();
+
+    const unsigned int my_rank    = Utilities::MPI::this_mpi_process(mpi_comm);
+    const unsigned int group_root = (my_rank / group_size) * group_size;
+
+    // create coarse grid on root processes
+    Triangulation<dim, spacedim> tria_serial;
+    if(my_rank == group_root)
+    {
+      auto [points, cell_data, sub_cell_data] = GridTools::get_coarse_mesh_description(temp_tria);
+
+      std::vector<std::pair<unsigned int, CellData<dim>>> cell_data_temp;
+
+      unsigned int counter = 0;
+
+      for(const auto & cell : temp_tria.cell_iterators_on_level(0))
+        cell_data_temp.emplace_back(cell->id().get_coarse_cell_id(), cell_data[counter++]);
+
+      std::sort(cell_data_temp.begin(), cell_data_temp.end(), [](const auto & a, const auto & b) {
+        return a.first < b.first;
+      });
+
+      cell_data.clear();
+
+      for(const auto & i : cell_data_temp)
+        cell_data.emplace_back(i.second);
+
+      tria_serial.create_triangulation(points, cell_data, sub_cell_data);
```

Comment:
This code only depends on `temp_tria`? -> implement as a free function to render code more readable/understandable?

Reply:
Good idea, I did this with the next block as well in a new function `replicate_distributed_triangulation_on_node`, to make that part clear from the name.